data/source/MongoDb.php

```diff
@@ -212,91 +233,98 @@ public static function enabled($feature = null) {
-	 *          `Mongo::setReadPreference()`. Defaults to null.
-	 *          Typically, these parameters are set in `Connections::add()`, when adding the
-	 *          adapter to the list of active connections.
 	 *
 	 *        Disables auto-connect, which is by default enabled in `Source`. Instead before
 	 *        each query execution the connection is checked and if needed (re-)established.
 	 * @return void
 	 */
 	public function __construct(array $config = []) {
-		if (class_exists($server = $this->_classes['server'], false)) {
-			$defaultHost = $server::DEFAULT_HOST . ':' . $server::DEFAULT_PORT;
-		} else {
-			$defaultHost = static::DEFAULT_HOST . ':' . static::DEFAULT_PORT;
-		}
 		$defaults = [
-			'host' => $defaultHost,
-			'login'      => null,
-			'password'   => null,
-			'database'   => null,
-			'timeout'    => 100,
-			'replicaSet' => false,
-			'schema'     => null,
-			'gridPrefix' => 'fs',
-			'w'          => 1,
-			'wTimeoutMS' => 10000,
-			'readPreference' => null,
-			'autoConnect' => false,
-			'dsn' => null
+			'host'               => static::DEFAULT_HOST . ':' . static::DEFAULT_PORT,
+			'login'              => null,
+			'password'           => null,
+			'database'           => null,
+			'dsn'                => null,
+			'timeout'            => 100,
+			'uriOptions'            => [
+				'w'                        => 'majority',
+				'wTimeoutMS'               => 10000,
+				'journal'                  => true,
+				'readConcernLevel'         => 'local',
+				'readPreference'           => 'primary',
+				'readPreferenceTags'       => [],
+			],
+			'driverOptions' => [
+				'allow_invalid_hostname' => false,
+				'weak_cert_validation'   => false
+			]
 		];
-		parent::__construct($config + $defaults);
+		$config = Set::merge($defaults, $config);
+		if (!isset($config['uriOptions']['connectTimeoutMS']) && $config['timeout']) {
+			$config['uriOptions']['connectTimeoutMS'] = $config['timeout'];
+		}
+		parent::__construct($config);
 	}
 
 	/**
 	 * Initializer. Adds operator handlers which will later allow to correctly cast any
-	 * values. Constructs a DSN from configuration.
+	 * values. Constructs a DSN from configuration, if not given.
 	 *
 	 * @see lithium\data\source\MongoDb::$_operators
 	 * @see lithium\data\source\MongoDb::_operators()
 	 * @return void
 	 */
 	protected function _init() {
-		$hosts = [];
-		foreach ((array) $this->_config['host'] as $host) {
-			$host = HostString::parse($host) + [
-				'host' => static::DEFAULT_HOST,
-				'port' => static::DEFAULT_PORT
-			];
-			$hosts[] = "{$host['host']}:{$host['port']}";
-		}
-		if ($this->_config['login']) {
-			$this->_config['dsn'] = sprintf(
-				'mongodb://%s:%s@%s/%s',
-				$this->_config['login'],
-				$this->_config['password'],
-				implode(',', $hosts),
-				$this->_config['database']
-			);
-		} else {
-			$this->_config['dsn'] = sprintf(
-				'mongodb://%s',
-				implode(',', $hosts)
-			);
-		}
 		parent::_init();
 
+		switch (true) {
+
```

Comment:
In this block a couple of blank lines sneaked in (below switch, throw and break).